### PR TITLE
Docs: How to use Node Material

### PR DESCRIPTION
**Pull request by sunag**
**Description**

I'm not an english expert, so it may need proofreading.

Link:
https://raw.githack.com/sunag/three.js/dev-nodes-doc/docs/manual/en/introduction/How-to-use-node-material.html

*This contribution is funded by [Google via Igalia](https://igalia.com)*

### CODE REVIEW

File: docs/list.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
 				"How to dispose of objects": "manual/en/introduction/How-to-dispose-of-objects",
 				"How to create VR content": "manual/en/introduction/How-to-create-VR-content",
 				"How to use post-processing": "manual/en/introduction/How-to-use-post-processing",
+				"How to use node material": "manual/en/introduction/How-to-use-node-material",
```
Comment (LeviPesin) [inline docs/list.json:25]: I would suggest to capitalize both words, i.e. write `How to use Node Material`.

**Comment by LeviPesin** [comment]
https://raw.githack.com/sunag/three.js/dev-nodes-doc/docs/manual/en/introduction/How-to-use-node-material.html

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 2]
```diff
@@ -0,0 +1,815 @@
...
+		</p>
+
+		<h2>How it works?</h2>
+
+		<p>
+		A `NodeMaterial` consists of `flows` which can have one or more `Nodes`, you can customize a `NodeMaterial` using existing `Nodes` and also create your own.
+		</p>
+
+		<p>
+		All `Nodes` can be connected to each other, and all must have automatic conversion according to the needs of their connections. Each Node and can vary between number of inputs and a single output with dynamic types.
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:23]: There is something wrong with the last phrase...

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+		`Nodes` can be values such as *Uniforms*, *Constants*, *Attributes of a Geometry* and *Mathematical Functions*.
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:24]: I would wrote something like "Nodes can be constants, uniforms, attributes, and functions" (not only mathematical).

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 4]
```diff
@@ -0,0 +1,815 @@
...
+		</p>
+
+		<p>
+		`Node` is responsible for emitting an snippet that can vary depending on the input arguments and output expected by the parent `Node`.
+		These small code represents the final result of `flow` of each property of a `NodeMaterial`. This code is generated from the `Node.generate()` function and changes according with the `Context` and `NodeBuilder` used.
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:29]: "These snippets represent the final code of the node."

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 5]
```diff
@@ -0,0 +1,815 @@
...
+		</code>
+
+		<h1>Shader Node</h1>
+		<p>
+		`NodeMaterial API` offers one own abstraction language so you don't have to rewrite `Shaders` for different outputs.
+		This also makes it easier to optimize the shaders themselves for the resources used in *three.js* since `NodeBuilder` can share the nodes.
+		</p>
+
+		<p>
+		The syntax is a combination of JS + GLSL, you can use JS to create the variables and flow while uses GLSL functions.
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:69]: "....while simutaneously using GLSL-written functions with them".

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 6]
```diff
@@ -0,0 +1,815 @@
...
+		</p>
+
+		<h2>Constants and explicit conversions</h2>
+
+		<p>
+		Input functions can be used to create code constants and explicit conversions.
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:75]: "...create contants and do explicit conversions".

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 7]
```diff
@@ -0,0 +1,815 @@
...
+		`NodeMaterial API` does all conversions automatically if the output and input are different.
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:76]: "...are of different types".

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 8]
```diff
@@ -0,0 +1,815 @@
...
+						<br>
+						[page:bool]( value )
+					</td>
+					<td>Returns a constant boolean value or convert a node to boolean.</td>
+				</tr>
+				<tr>
+					<td>
+						[page:color]( [param:Number value] )
+						<br>
+						[page:color]( [param:Number r], [param:Number g], [param:Number b] )
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:99]: I think signatures like `bool( Node value )` and `color( Node value, Node value, Node value )` also should be added here.

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 9]
```diff
@@ -0,0 +1,815 @@
...
+const node = mat4();
+
+// using static index
+const nodeColumn0 = node[ 0 ]; // only primitive value
+		</code>
+
+		<h4>Example</h4>
+		<code>
+const node = mat4();
+const index = uint( 0 );
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:247]: I would prefer to have here something more complex... E.g. `index = add(2, 2)`?

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 10]
```diff
@@ -0,0 +1,815 @@
...
+				<tr>
+					<th>Function</th>
+					<th>Description</th>
+					<th>Output</th>
+				</tr>
+			</thead>
+			<tbody>
+				<tr>
+					<td>[page:uv]( [param:Number index] )</td>
+					<td>Return the uv by index.</td>
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:487]: "the uv-coordinates with that index".

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 11]
```diff
@@ -0,0 +1,815 @@
...
+					[page:normalGeometry]
+					<br>
+					[page:normalLocal]
+					<br>
+					[page:normalWorld]
+					<br>
+					[page:normalView]
+					<br>
+					</td>
+					<td>Return the normal from the coordinate space chosen.</td>
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:506]: "the normal vector"

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 12]
```diff
@@ -0,0 +1,815 @@
...
+					[page:positionLocal]
+					<br>
+					[page:positionWorld]
+					<br>
+					[page:positionView]
+					<br>
+					[page:positionViewDirection]
+					<br>
+					</td>
+					<td>Return the position from the coordinate space chosen.</td>
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:522]: `the position in the coordinate space chosen"

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 13]
```diff
@@ -0,0 +1,815 @@
...
+			</thead>
+			<tbody>
+				<tr>
+					<td>[page:cameraPosition]</td>
+					<td>Returns a vector representing the position of the object in world space.</td>
+					<td>vec3</td>
+				</tr>
+				<tr>
+					<td>[page:cameraProjectionMatrix]</td>
+					<td>This is the matrix which contains the projection.</td>
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:558]: "Return the matrix"

Comment (LeviPesin) [reply docs/manual/en/introduction/How-to-use-node-material.html:558]: And the same for two entries below.

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 14]
```diff
@@ -0,0 +1,815 @@
...
+				<tr>
+					<th>Function</th>
+					<th>Description</th>
+					<th>Output</th>
+				</tr>
+			</thead>
+			<tbody>
+				<tr>
+					<td>[page:texture]( [param:Texture texture], uv, level )</td>
+					<td>Return the addition of two or more value.</td>
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:670]: Incorrect description here and in the next entry.

File: docs/manual/en/introduction/How-to-use-node-material.html [hunk 15]
```diff
@@ -0,0 +1,815 @@
...
+			<thead>
+				<tr>
+					<th>Function</th>
+					<th>Description</th>
+				</tr>
+			</thead>
+			<tbody>
+				<tr>
+					<td>[page:context]( node, [param:Object context] )</td>
+					<td>Return the value of the node parameter.</td>
```
Comment (LeviPesin) [inline docs/manual/en/introduction/How-to-use-node-material.html:751]: "...the value of the node parameter, but in the new context".

**Comment by LeviPesin** [comment]
I think there is too much of \`\` used on the page...

**Comment by hybridherbst** [comment]
One question that comes up for me whenever looking at the state of Node Material: is there a way to use it with GltfLoader and/or is there already feature parity to MeshStandardMaterial in that regard? Would be nice to see this covered in the docs.

**Comment by DennisSmolek** [comment]
This thread is pretty old, I'm wondering what the status is of the docs and how up to date the docs are with the current state of the node system. 
If you need someone to go over them for english/editing I'm willing. I'm also willing to update them etc.

**Comment by sunag** [comment]
This is being updated here https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language

**Pull request closed without merging**